### PR TITLE
[Feature] Standardize Router Model dataset, training-run, and artifact manifests

### PR DESCRIPTION
**Pull request by subin9**
Related #3199

## Purpose

Nothing in the model path pins a revision. `tools/make/models.mk` downloads with `hf download $(HF_ORG)/$$model` and no `--revision`, so `models/` holds whatever `main` pointed at that day, and no artifact records which run produced it. The prompt-guard LoRA script seeded `train_test_split` with `random_state=42` but never seeded torch or the classifier head.

This adds four manifest kinds, dataset, run, artifact and evaluation, as versioned YAML with JSON Schema, following `router_calibration_schema.py`: exact `schema_version`, unknown-field rejection, and no YAML anchors, aliases, tags or merge keys.

Schema validation proves each manifest is well formed. It does not prove the four describe the same model, so cross-reference validation fails on a revision that is not a commit sha, a label mapping that is not contiguous `0..n-1`, an `identity.digest` that disagrees with the file list it summarises, a reference that resolves to nothing, an evaluation reporting more rows than the split holds or fewer with no declared `sample_limit`, and a label mapping that differs between any two manifests.

The label check is the one that motivated this. `constants.py` lists the intent classifier's 14 labels in a different order than the artifact's own `label_mapping.json`, which costs 27 accuracy points on MMLU-Pro test, 0.9445 down to 0.6746. Nothing in the metrics looks wrong, so only a cross-reference catches it.

The dataset schema takes a `composite` source type, since the prompt-guard training set is built in code from `lmsys/toxic-chat`, `OpenSafetyLab/Salad-Data` and the in-repo pattern lists. Each upstream stays pinned separately, so a mixture is traceable without hosting it.

`prompt_guard` is the one workflow wired up. `save_training_artifacts` now writes the dataset, run and artifact manifests alongside the adapter, and `--seed` calls `set_seed` before any sampling so the recorded seed describes the run. Emission raises rather than writing a partial bundle.

The evaluation manifest is emitted by whatever measures the artifact, which is the baseline harness in the #3197 branch. A bundle without one does not validate, since an artifact with no measured evaluation is not qualified.

Non-goals from the issue hold: no other training script changes, nothing is hosted in the repo, and nothing promotes an artifact.

## Test Plan

- Unit tests for the schema, cross-reference, redaction and emission behaviour.
- One real training run that emits the manifests, then validate the bundle.
- The changed-file lint and structure gates.

## Test Result

```
pytest src/training/model_eval/test_provenance.py \
       src/training/model_classifier/prompt_guard_fine_tuning_lora/
26 passed

ruff check --config tools/linter/python/.ruff.toml <changed files>   All checks passed
tools/agent/scripts/structure_check.py <changed files>               no errors
codespell <changed files>                                            clean
black --check <changed files>                                        clean
```

One training run on the jailbreak mixture, `--model mmbert-base --epochs 1 --max-samples 200 --seed 42`, then:

```
python -m provenance.cli validate <run>/manifests
ok dataset    prompt-guard-mixture-max200
ok run        prompt-guard-mmbert-base-r8-seed42
ok artifact   prompt-guard-mmbert-base-r8-lora
```

The bundle is reported incomplete until an evaluation manifest joins it, which is the intended behaviour and what the #3197 branch adds.

Permuting the class order in one manifest fails the bundle. So does a digest that disagrees with the file list, and a reference that resolves to nothing.

Remaining: `structure_check.py` still warns that `jailbreak_bert_finetuning_lora.py` is long. It was 764 lines before this change and is 798 now, under the 800 limit. Putting the emission in `save_training_artifacts` rather than in `main()` is what keeps it there, and it is also where the rest of the run's output is written.

No CI job runs the tests under `src/training/`, so the commands above are local. Wiring that up looks like a separate change.

### CODE REVIEW

**Review by FAUST-BENCHOU — COMMENTED** [review]
Since the manifest format is intended to standardize Router Models, should we define a common emission interface for other training scripts instead of wiring only prompt_guard here? Otherwise provenance coverage may depend on individual training workflows.

**Comment by Xunzhuo** [comment]
Thanks for putting this together. #3199 is still awaiting acceptance and assignment, so I’m holding the implementation review for now. Please pause here until the Router Models maintainers accept the bounded contract; after that, rebase and we’ll review the exact manifest surface.

**Comment by subin9** [comment]
@FAUST-BENCHOU The shared surface is already there: `src/training/model_eval/provenance/emit.py` holds the builders, the revision and digest resolution, and `write_manifest`, which validates before it returns. `jailbreak_provenance.py` is a thin adapter over it.

What is left per script is the part that genuinely differs: which upstreams the dataset is built from, and what the hyperparameters are called. Wiring a second script means calling the same builders with its own values.

Agreed that coverage should not depend on each workflow remembering to do it. #3200 is where that becomes enforceable, since it requires one model to complete the path from declared inputs to a promoted artifact. Wiring the rest here would have gone against the non-goal in #3199 of rewriting every training script in one change, but happy to add a second script if you would rather see the interface exercised twice before it settles.

Thank you :)

@Xunzhuo Understood, pausing here for now. Thank you!

**Comment by subin9** [comment]
#3199 accepted, so unpaused for review :)

**Review by Xunzhuo — CHANGES_REQUESTED** [review]
Thanks for wiring provenance into the training lane. The manifest is not bound to the bytes actually trained: the model and datasets load without revisions, then the current Hub SHAs are resolved only after training. A moving ref or stale cache can therefore produce valid-looking false provenance. Please resolve and pin the exact revisions into the actual loads. The workflow also emits only dataset, run, and artifact manifests; it omits the evaluation manifest required for #3199 completion, so please wire that producer before closing the issue.

**Comment by subin9** [comment]
Revisions are resolved once before anything loads and passed into the loads themselves, so the datasets and the base model are read at the commit the manifests name. Emission records those values instead of resolving again, and a test swaps the resolver for one that raises to keep it that way.

The workflow writes the evaluation manifest now: the adapter is measured on the validation split it held out, with the metrics, calibration, abstention curve and latency the schema requires, and split_rule: by_row, since this workflow splits its own mixture positionally. validate_bundle over the four manifests is a test.

**Review by Xunzhuo — APPROVED** [review]
